## Preprocessing

In [1]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd 
application_df = pd.read_csv("../Resources/charity_data.csv")
application_df

EIN                                               NAME  \
0       10520599                       BLUE KNIGHTS MOTORCYCLE CLUB   
1       10531628             AMERICAN CHESAPEAKE CLUB CHARITABLE TR   
2       10547893                 ST CLOUD PROFESSIONAL FIREFIGHTERS   
3       10553066                     SOUTHSIDE ATHLETIC ASSOCIATION   
4       10556103           GENETIC RESEARCH INSTITUTE OF THE DESERT   
...          ...                                                ...   
34294  996009318              THE LIONS CLUB OF HONOLULU KAMEHAMEHA   
34295  996010315           INTERNATIONAL ASSOCIATION OF LIONS CLUBS   
34296  996012607                                PTA HAWAII CONGRESS   
34297  996015768  AMERICAN FEDERATION OF GOVERNMENT EMPLOYEES LO...   
34298  996086871                           WATERHOUSE CHARITABLE TR   

      APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0                  T10       Independent          C1000    ProductDev   
1                   T3       Independent          C2000  Preservation   
2                   T5  CompanySponsored          C3000    ProductDev   
3                   T3  CompanySponsored          C2000  Preservation   
4                   T3       Independent          C1000     Heathcare   
...                ...               ...            ...           ...   
34294               T4       Independent          C1000    ProductDev   
34295               T4  CompanySponsored          C3000    ProductDev   
34296               T3  CompanySponsored          C2000  Preservation   
34297               T5       Independent          C3000    ProductDev   
34298               T3       Independent          C1000  Preservation   

       ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS   ASK_AMT  \
0       Association       1              0                      N      5000   
1      Co-operative       1         1-9999                      N    108590   
2       Association       1              0                      N      5000   
3             Trust       1    10000-24999                      N      6692   
4             Trust       1  100000-499999                      N    142590   
...             ...     ...            ...                    ...       ...   
34294   Association       1              0                      N      5000   
34295   Association       1              0                      N      5000   
34296   Association       1              0                      N      5000   
34297   Association       1              0                      N      5000   
34298  Co-operative       1          1M-5M                      N  36500179   

       IS_SUCCESSFUL  
0                  1  
1                  1  
2                  0  
3                  1  
4                  1  
...              ...  
34294              0  
34295              0  
34296              0  
34297              1  
34298              0  

[34299 rows x 12 columns]

In [2]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
#  YOUR CODE GOES HERE
application_df = application_df.drop('EIN', axis=1)
application_df = application_df.drop('NAME', axis=1)
application_df.head(5)

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0              T10       Independent          C1000    ProductDev   
1               T3       Independent          C2000  Preservation   
2               T5  CompanySponsored          C3000    ProductDev   
3               T3  CompanySponsored          C2000  Preservation   
4               T3       Independent          C1000     Heathcare   

   ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  \
0   Association       1              0                      N     5000   
1  Co-operative       1         1-9999                      N   108590   
2   Association       1              0                      N     5000   
3         Trust       1    10000-24999                      N     6692   
4         Trust       1  100000-499999                      N   142590   

   IS_SUCCESSFUL  
0              1  
1              1  
2              0  
3              1  
4              1

In [3]:
# Determine the number of unique values in each column.
#  YOUR CODE GOES HERE
uniques_columns = application_df.nunique()
uniques_columns

APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
USE_CASE                     5
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64

In [4]:
# Look at APPLICATION_TYPE value counts for binning
#  YOUR CODE GOES HERE
app_unique = application_df["APPLICATION_TYPE"].value_counts()
app_unique

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [5]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
#  YOUR CODE GOES HERE
application_types_to_replace = ['T9', 'T13', 'T12', 'T2', 'T25', 'T14', 'T29', 'T15', 'T17']


# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
application_df['APPLICATION_TYPE'].value_counts()

T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: APPLICATION_TYPE, dtype: int64

In [6]:
# Look at CLASSIFICATION value counts for binning
#  YOUR CODE GOES HERE
class_unique = application_df["CLASSIFICATION"].value_counts()
class_unique

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C2150        1
C2500        1
C1900        1
C1248        1
C6100        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [7]:
# You may find it helpful to look at CLASSIFICATION value counts >1
#  YOUR CODE GOES HERE
class_unique_nonzero= class_unique[class_unique>1]
class_unique_small = class_unique[class_unique < 1883]
df_class = pd.DataFrame(class_unique_small)
classifications_to_replace = df_class.index
classifications_to_replace

Index(['C7000', 'C1700', 'C4000', 'C5000', 'C1270', 'C2700', 'C2800', 'C7100',
       'C1300', 'C1280', 'C1230', 'C1400', 'C2300', 'C7200', 'C1240', 'C8000',
       'C7120', 'C1500', 'C6000', 'C1800', 'C1250', 'C8200', 'C1238', 'C1278',
       'C1237', 'C1235', 'C7210', 'C4100', 'C1720', 'C2400', 'C1600', 'C1257',
       'C1260', 'C2710', 'C0', 'C3200', 'C1267', 'C1256', 'C1234', 'C1246',
       'C1580', 'C1236', 'C2561', 'C1570', 'C5200', 'C2600', 'C1245', 'C3700',
       'C4200', 'C1283', 'C2190', 'C1820', 'C8210', 'C2570', 'C1728', 'C1732',
       'C2170', 'C2380', 'C4500', 'C1370', 'C4120', 'C2150', 'C2500', 'C1900',
       'C1248', 'C6100'],
      dtype='object')

In [8]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
#  YOUR CODE GOES HERE

# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")
    
# Check to make sure binning was successful
application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
Other     2261
C3000     1918
C2100     1883
Name: CLASSIFICATION, dtype: int64

In [9]:
# Convert categorical data to numeric with `pd.get_dummies`
#  YOUR CODE GOES HERE
with_dummies = pd.get_dummies(application_df)
with_dummies.head()

STATUS  ASK_AMT  IS_SUCCESSFUL  APPLICATION_TYPE_Other  \
0       1     5000              1                       0   
1       1   108590              1                       0   
2       1     5000              0                       0   
3       1     6692              1                       0   
4       1   142590              1                       0   

   APPLICATION_TYPE_T10  APPLICATION_TYPE_T19  APPLICATION_TYPE_T3  \
0                     1                     0                    0   
1                     0                     0                    1   
2                     0                     0                    0   
3                     0                     0                    1   
4                     0                     0                    1   

   APPLICATION_TYPE_T4  APPLICATION_TYPE_T5  APPLICATION_TYPE_T6  ...  \
0                    0                    0                    0  ...   
1                    0                    0                    0  ...   
2                    0                    1                    0  ...   
3                    0                    0                    0  ...   
4                    0                    0                    0  ...   

   INCOME_AMT_1-9999  INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  \
0                  0                       0                         0   
1                  1                       0                         0   
2                  0                       0                         0   
3                  0                       1                         0   
4                  0                       0                         1   

   INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  \
0                   0                 0                       0   
1                   0                 0                       0   
2                   0                 0                       0   
3                   0                 0                       0   
4                   0                 0                       0   

   INCOME_AMT_50M+  INCOME_AMT_5M-10M  SPECIAL_CONSIDERATIONS_N  \
0                0                  0                         1   
1                0                  0                         1   
2                0                  0                         1   
3                0                  0                         1   
4                0                  0                         1   

   SPECIAL_CONSIDERATIONS_Y  
0                         0  
1                         0  
2                         0  
3                         0  
4                         0  

[5 rows x 44 columns]

In [10]:
# Split our preprocessed data into our features and target arrays
#  YOUR CODE GOES HERE
y = with_dummies['IS_SUCCESSFUL']
X = with_dummies.drop('IS_SUCCESSFUL', axis=1)

# Split the preprocessed data into a training and testing dataset
#  YOUR CODE GOES HERE
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

In [11]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Compile, Train and Evaluate the Model

In [12]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
#  YOUR CODE GOES HERE

nn = tf.keras.models.Sequential()

nn.add(tf.keras.layers.Dense(units=500, activation="relu", input_dim=43))
nn.add(tf.keras.layers.Dense(units=500, activation="sigmoid"))
nn.add(tf.keras.layers.Dense(units=500, activation="relu"))
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 500)               22000     
_________________________________________________________________
dense_1 (Dense)              (None, 500)               250500    
_________________________________________________________________
dense_2 (Dense)              (None, 500)               250500    
_________________________________________________________________
dense_3 (Dense)              (None, 1)                 501       
Total params: 523,501
Trainable params: 523,501
Non-trainable params: 0
_________________________________________________________________


In [13]:
# Compile the model
#  YOUR CODE GOES HERE
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [14]:
# Train the model
#  YOUR CODE GOES HERE
fit_model = nn.fit(X_train_scaled, y_train, epochs=10)

Epoch 1/10
804/804 [==============================] - 8s 9ms/step - loss: 0.5756 - accuracy: 0.7164
Epoch 2/10
804/804 [==============================] - 7s 9ms/step - loss: 0.5577 - accuracy: 0.7293: 0s - loss: 0.5
Epoch 3/10
804/804 [==============================] - 8s 10ms/step - loss: 0.5533 - accuracy: 0.7304
Epoch 4/10
804/804 [==============================] - 8s 10ms/step - loss: 0.5514 - accuracy: 0.7311
Epoch 5/10
804/804 [==============================] - 8s 10ms/step - loss: 0.5500 - accuracy: 0.7317
Epoch 6/10
804/804 [==============================] - 7s 9ms/step - loss: 0.5488 - accuracy: 0.7324
Epoch 7/10
804/804 [==============================] - 7s 9ms/step - loss: 0.5480 - accuracy: 0.7331
Epoch 8/10
804/804 [==============================] - 7s 9ms/step - loss: 0.5468 - accuracy: 0.7339
Epoch 9/10
804/804 [==============================] - 8s 9ms/step - loss: 0.5461 - accuracy: 0.7339
Epoch 10/10
804/804 [==============================] - 7s 9ms/step - loss: 0.5456

In [15]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 2s - loss: 0.5528 - accuracy: 0.7303
Loss: 0.5527547597885132, Accuracy: 0.7302623987197876


In [16]:
# Export our model to HDF5 file
#  YOUR CODE 
nn.save('model_homework.h5')

Overview: The goal of the homework was finding the optimal model to predict the likelihood of success of charities with the inputs
Results: Got a model with accuracy of 75%
  * Data Preprocessing
    * What variable(s) are considered the target(s) for your model? 
        * The variable considered the target for this model is the success of the charity (IS_SUCCESSFUL)
    * What variable(s) are considered to be the features for your model?
        * APPLICATION_TYPE
        * AFFILIATION CLASSIFICATION
        * USE_CASE
        * ORGANIZATION STATUS
        * INCOME_AMT
        * SPECIAL_CONSIDERATIONS
        * ASK_AMT
    * What variable(s) are neither targets nor features, and should be removed from the input data?
        * EIN
        * NAME
  * Compiling, Training, and Evaluating the Model
    * How many neurons, layers, and activation functions did you select for your neural network model, and why?
        * There are four layers and 500 neurons to try to improve the accuracy of the neural network
    * Were you able to achieve the target model performance?
        *  I was not able to get to an accuracy of 75%. The highest I got was around 73%
    * What steps did you take to try and increase model performance?
 Summary: Found a model that took the inputs but only got a model of around 73%